drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <net/socket.h>
+#include <init.h>
+#include <bsd_limits.h>
+
+#include <at_cmd.h>
+
+LOG_MODULE_REGISTER(at_cmd, CONFIG_AT_CMD_LOG_LEVEL);
+
+#define RESPONSE_MAX_LEN  CONFIG_AT_CMD_RESPONSE_MAX_LEN
+#define THREAD_PRIORITY   K_PRIO_PREEMPT(CONFIG_AT_CMD_THREAD_PRIO)
+
+#define AT_CMD_OK_STR    "OK"
+#define AT_CMD_ERROR_STR "ERROR"
+#define AT_CMD_CMS_STR   "+CMS:"
+#define AT_CMD_CME_STR   "+CME:"
+
+static K_THREAD_STACK_DEFINE(socket_thread_stack, CONFIG_AT_CMD_THREAD_STACK_SIZE);
+static K_MUTEX_DEFINE(cmd_pending);
+
+K_MSGQ_DEFINE(return_code_msq, sizeof(enum at_cmd_rc), 2, 4);
+
+static int              common_socket_fd;
+static int              last_error;
+static char             *response_buf;
+static u32_t            response_buf_len;
+
+static struct k_thread  socket_thread;
+static at_cmd_handler_t notification_handler;
+static at_cmd_handler_t current_cmd_handler;
+
+struct callback_work_item {
+	struct k_work    work;
+	char             data[RESPONSE_MAX_LEN];
+	at_cmd_handler_t callback;
+};
+
+static int open_socket(void)
+{
+	common_socket_fd = socket(AF_LTE, 0, NPROTO_AT);
+
+	if (common_socket_fd == -1) {
+		LOG_ERR("AT socket creation failed");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static enum at_cmd_rc get_return_code(char *buf)
+{
+	enum at_cmd_rc return_code = AT_CMD_NOTIFICATION;
+	char *tmpstr = NULL;
+
+	do {
+		tmpstr = strstr(buf, AT_CMD_OK_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_OK;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_ERROR_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CMS_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR_CMS;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CME_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR_CME;
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_ERROR_CMS) ||
+	    (return_code == AT_CMD_ERROR_CME)) {
+		last_error = atoi(&buf[ARRAY_SIZE(AT_CMD_CMS_STR) - 1]);
+	} else {
+		last_error = 0;
```

Comment:
If you get only "ERROR" (when CME and CMS are disabled), will `last_error` will be `0`?

Reply:
This one is a bit tricky because when ERROR is returned there is no errorcode, but I don't think keeping the error code from the last CME or CMS (if there has been any) will be correct either. Maybe a generic -1? CME and CMS will not return -1 and then we leave 0 for when there has not been any error of any kind.

Reply:
This is related to my other comment. I think we should return two information: the state, to indicates if it was a generic error, CME error or CMS error depending on the socket configuration, and the error code. For all states, error codes have different meanings and ranges.